src/simulators.jl

```diff
@@ -438,3 +438,110 @@ function B_step!(s, dt_eff, acceleration_vector, compute_forces::Bool, n_threads
     end
     s.velocities += dt_eff * remove_molar.(acceleration_vector)
 end
+
+"""
+    TemperatureREMD(; <keyword arguments>)
+
+A simulator for a parallel temperature replica exchange (TREX) simulation on a [`ReplicaSystem`](@ref). More information on this algorithm can be found in [Sugita Y., Okamoto Y. 1999](https://doi.org/10.1016/S0009-2614(99)01123-9).
+The corresponding [`ReplicaSystem`](@ref) should have the same number of replicas as the number of temperatures in the simulator.
+
+arguments:
+- `dt::S`: the time step of the simulation.
+- `temperatures::TP`: the temperatures corresponding to the replicas.
+- `simulators::ST`: individual simulators for simulating each replica.
+- `exchange_time::ET`: the time interval between replica exchange attempt.
+"""
+struct TemperatureREMD{N, T, S, DT, TP, ST, ET}
+    dt::DT
+    temperatures::TP
+    simulators::ST
+    exchange_time::ET
+end
+
+function TemperatureREMD(;
+                dt,
+                temperatures,
+                simulators,
+                exchange_time,
+                kwargs...)
+    S = eltype(simulators)
+    T = eltype(temperatures)
+    N = length(temperatures)
+    DT = typeof(dt)
+    TP = typeof(temperatures)
+    ET = typeof(exchange_time)
+    if length(simulators) != length(temperatures)
+        throw(ArgumentError("Number of temperatures must match number of simulators"))
+    end
+    if exchange_time <= dt
+        throw(ArgumentError("Exchange time must be greater than the time step"))
+    end
+    simulators = Tuple(simulators[i] for i in 1:N)
+    ST = typeof(simulators)
+    
+    return TemperatureREMD{N, T, S, DT, TP, ST, ET}(temperatures, simulators, exchange_time, dt)
+end
+
+function simulate!(sys::ReplicaSystem,
+                    sim::TemperatureREMD,
+                    n_steps::Int;
+                    n_threads::Int=Threads.nthreads())
+    if sys.n_replicas != length(sim.simulators)
+        error("Number of replicas in ReplicaSystem and simulators in TemperatureREMD must match.")
+    end
+
+    if n_threads > sys.n_replicas
+        thread_div = equal_parts(n_threads, sys.n_replicas)
+    else # pass 1 thread per replica
+        thread_div = equal_parts(sys.n_replicas, sys.n_replicas)
+    end
```

Comment:
A static division of threads is done among all the replicas. I am not sure whether this can be made dynamic or not without much overhaul.

Reply:
A static division is probably okay for now.

Yes CellListMap.jl uses a number of threads up to `Threads.nthreads()` when `parallel=true`, I opened an issue to discuss this (https://github.com/m3g/CellListMap.jl/issues/64).

Reply:
Looks like the response has the answer and we can use something like `nbatches=(min(n, 8), n)` where `n` is the threads per replica.

Reply:
There is one problem though, the `nbatches` argument is passed to the constructor while defining `CellListNeighborFinder`, which is then used in the underlying cell list type. So we can not update it later say during the thread division in the replica exchange simulator.

Reply:
In that case it might be easier to leave it up to the user to specify when giving the neighbour finder to `ReplicaSystem`.